datajunction-server/datajunction_server/sql/dag.py

```diff
@@ -226,6 +227,202 @@ async def get_upstream_nodes(
+                        [node.name] if dimension_node.name != node.name else []
+                    ) + [dimensions_map[int(node_id)].name for node_id in path]
+                    if ref_link := await build_reference_link(
+                        session,
+                        col,
+                        join_path,
+                        role,
+                    ):
+                        reference_links.append(ref_link)
+
+    # Build all dimension attributes from the dimension nodes in the graph
+    graph_dimensions = [
+        DimensionAttributeOutput(
+            name=f"{dim.name}.{col.name}" + (f"[{'->'.join(role)}]" if role else ""),
+            node_name=dim.name,
+            node_display_name=dim.current.display_name,
+            properties=col.attribute_names(),
+            type=str(col.type),
+            path=[node_name] + [dimensions_map[int(node_id)].name for node_id in path],
+        )
+        for dim, path, role in dimension_nodes_and_paths
+        for col in dim.current.columns
+    ]
+
+    # Build all local dimension attributes from the original node
+    local_dimensions = [
+        DimensionAttributeOutput(
+            name=f"{node.name}.{col.name}",
+            node_name=node.name,
+            node_display_name=node.current.display_name,
+            properties=col.attribute_names(),
+            type=str(col.type),
+            path=[],
+        )
+        for col in node.current.columns
+    ]
+    local_dimensions = [
+        dim
+        for dim in local_dimensions
+        if "primary_key" in (dim.properties or [])
+        or "dimension" in (dim.properties or [])
+        or node.type == NodeType.DIMENSION
+    ]
+    return reference_links + graph_dimensions + local_dimensions
+
+
+async def get_dimension_nodes(
+    session: AsyncSession,
+    node: Node,
+    include_deactivated: bool = True,
+) -> list[tuple[Node, list[str], list[str] | None]]:
+    """
+    Discovers all dimension nodes in the given node's dimensions graph using a recursive
+    CTE query to build out the dimension links.
+    """
+    dag = (
+        (
+            select(
+                DimensionLink.node_revision_id,
+                NodeRevision.id,
+                NodeRevision.node_id,
+                array([NodeRevision.node_id]).label("join_path"),  # start path
+                array([DimensionLink.role]).label("role"),
+            )
+            .where(DimensionLink.node_revision_id == node.current.id)
+            .join(Node, DimensionLink.dimension_id == Node.id)
+            .join(
+                NodeRevision,
+                (Node.id == NodeRevision.node_id)
+                & (Node.current_version == NodeRevision.version),
+            )
+        )
+        .cte("dimensions", recursive=True)
+        .suffix_with(
+            "CYCLE node_id SET is_cycle USING path",
+        )
+    )
+
+    paths = dag.union_all(
+        select(
+            dag.c.node_revision_id,
```

Comment:
What is `.c` ?

Reply:
It's how you reference a column on a SQLAlchemy query or table object

Reply:
Wow. What a shortcut... Show how much I know ORMs :)